File: android/src/main/java/expo/modules/nuxieexpo/bridges/NuxiePurchaseDelegateBridge.kt
package expo.modules.nuxieexpo.bridges

import ai.nuxie.sdk.billing.NuxiePurchaseDelegate
import ai.nuxie.sdk.billing.PurchaseResult
import ai.nuxie.sdk.billing.RestoreResult
import ai.nuxie.sdk.billing.StoreProduct
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.withTimeoutOrNull

class NuxiePurchaseDelegateBridge(
  private val emitEvent: (eventName: String, payload: Map<String, Any?>) -> Unit,
  private val timeoutMs: Long = 60_000,
) : NuxiePurchaseDelegate {
  private val purchaseRequests = ConcurrentHashMap<String, CompletableDeferred<PurchaseResult>>()
  private val restoreRequests = ConcurrentHashMap<String, CompletableDeferred<RestoreResult>>()

  override suspend fun purchase(product: StoreProduct): PurchaseResult {
    val requestId = UUID.randomUUID().toString()
    val deferred = CompletableDeferred<PurchaseResult>()
    purchaseRequests[requestId] = deferred

    emitEvent(
      "onPurchaseRequest",
      mapOf(
        "request_id" to requestId,
        "platform" to "android",
        "product_id" to product.productId,
        "store_product_id" to product.storeProductId,
        "base_plan_id" to product.basePlanId,
        "purchase_option_id" to product.purchaseOptionId,
        "offer_id" to product.offerId,
        "placement_id" to product.placementId,
        "display_name" to product.rawProduct?.name,
        "display_price" to null,
        "timestamp_ms" to System.currentTimeMillis(),
      ),
    )

    return try {
      withTimeoutOrNull(timeoutMs) { deferred.await() }
        ?: PurchaseResult.Failed(bridgeError("purchase_timeout"))
    } finally {
      purchaseRequests.remove(requestId)
    }
  }

  override suspend fun restorePurchases(): RestoreResult {
    val requestId = UUID.randomUUID().toString()
    val deferred = CompletableDeferred<RestoreResult>()
    restoreRequests[requestId] = deferred

    emitEvent(
      "onRestoreRequest",
      mapOf(
        "request_id" to requestId,
        "platform" to "android",
        "timestamp_ms" to System.currentTimeMillis(),
      ),
    )

    return try {
      withTimeoutOrNull(timeoutMs) { deferred.await() }
        ?: RestoreResult.Failed(bridgeError("restore_timeout"))
    } finally {
      restoreRequests.remove(requestId)
    }
  }

  fun completePurchase(requestId: String, payload: Map<String, Any?>) {
    purchaseRequests.remove(requestId)?.complete(purchaseResult(payload))
  }

  fun completeRestore(requestId: String, payload: Map<String, Any?>) {
    restoreRequests.remove(requestId)?.complete(restoreResult(payload))
  }

  fun cancelPending(reason: String) {
    purchaseRequests.values.forEach { request ->
      request.complete(PurchaseResult.Failed(bridgeError(reason)))
    }
    restoreRequests.values.forEach { request ->
      request.complete(RestoreResult.Failed(bridgeError(reason)))
    }
    purchaseRequests.clear()
    restoreRequests.clear()
  }

  private fun purchaseResult(payload: Map<String, Any?>): PurchaseResult =
    when ((payload["type"] as? String)?.lowercase()) {
      "purchased" -> PurchaseResult.Purchased
      "cancelled" -> PurchaseResult.Cancelled
      "pending" -> PurchaseResult.Pending
      else -> PurchaseResult.Failed(
        bridgeError((payload["message"] as? String) ?: "purchase_failed"),
      )
    }

  private fun restoreResult(payload: Map<String, Any?>): RestoreResult =
    when ((payload["type"] as? String)?.lowercase()) {
      "restored" -> RestoreResult.Restored
      "no_purchases" -> RestoreResult.NoPurchases
      else -> RestoreResult.Failed(
        bridgeError((payload["message"] as? String) ?: "restore_failed"),
      )
    }

  private fun bridgeError(message: String): Throwable = IllegalStateException(message)
}

File: ios/Bridges/NuxiePurchaseDelegateBridge.swift
import Foundation

#if canImport(Nuxie)
import Nuxie

final class NuxiePurchaseDelegateBridge: NuxiePurchaseDelegate, @unchecked Sendable {
  private let emit: (String, [String: Any]) -> Void
  private let timeoutSeconds: TimeInterval
  private let lock = NSLock()
  private var purchaseContinuations: [String: CheckedContinuation<PurchaseResult, Never>] = [:]
  private var restoreContinuations: [String: CheckedContinuation<RestoreResult, Never>] = [:]

  init(
    timeoutSeconds: TimeInterval = 60,
    emit: @escaping (String, [String: Any]) -> Void
  ) {
    self.timeoutSeconds = timeoutSeconds
    self.emit = emit
  }

  func purchase(product: StoreProduct) async -> PurchaseResult {
    let requestId = UUID().uuidString
    let payload: [String: Any] = [
      "request_id": requestId,
      "platform": "ios",
      "product_id": product.productId,
      "store_product_id": product.storeProductId,
      "base_plan_id": NSNull(),
      "purchase_option_id": NSNull(),
      "offer_id": NSNull(),
      "placement_id": product.placementId,
      "display_name": product.name,
      "display_price": product.price,
      "timestamp_ms": Int(Date().timeIntervalSince1970 * 1_000),
    ]

    return await withCheckedContinuation { continuation in
      lock.withLock {
        purchaseContinuations[requestId] = continuation
      }
      emit("onPurchaseRequest", payload)
      schedulePurchaseTimeout(requestId: requestId)
    }
  }

  func restorePurchases() async -> RestoreResult {
    let requestId = UUID().uuidString
    let payload: [String: Any] = [
      "request_id": requestId,
      "platform": "ios",
      "timestamp_ms": Int(Date().timeIntervalSince1970 * 1_000),
    ]

    return await withCheckedContinuation { continuation in
      lock.withLock {
        restoreContinuations[requestId] = continuation
      }
      emit("onRestoreRequest", payload)
      scheduleRestoreTimeout(requestId: requestId)
    }
  }

  func completePurchase(requestId: String, payload: [String: Any]) {
    let continuation = lock.withLock {
      purchaseContinuations.removeValue(forKey: requestId)
    }
    continuation?.resume(returning: purchaseResult(from: payload))
  }

  func completeRestore(requestId: String, payload: [String: Any]) {
    let continuation = lock.withLock {
      restoreContinuations.removeValue(forKey: requestId)
    }
    continuation?.resume(returning: restoreResult(from: payload))
  }

  func cancelPending() {
    let (purchases, restores) = lock.withLock {
      let purchases = Array(purchaseContinuations.values)
      let restores = Array(restoreContinuations.values)
      purchaseContinuations.removeAll()
      restoreContinuations.removeAll()
      return (purchases, restores)
    }
    purchases.forEach { $0.resume(returning: .failed(bridgeError("sdk_shutdown"))) }
    restores.forEach { $0.resume(returning: .failed(bridgeError("sdk_shutdown"))) }
  }

  private func schedulePurchaseTimeout(requestId: String) {
    Task { [weak self] in
      guard let self else { return }
      try? await Task.sleep(nanoseconds: UInt64(self.timeoutSeconds * 1_000_000_000))
      let continuation = self.lock.withLock {
        self.purchaseContinuations.removeValue(forKey: requestId)
      }
      continuation?.resume(returning: .failed(self.bridgeError("purchase_timeout")))
    }
  }

  private func scheduleRestoreTimeout(requestId: String) {
    Task { [weak self] in
      guard let self else { return }
      try? await Task.sleep(nanoseconds: UInt64(self.timeoutSeconds * 1_000_000_000))
      let continuation = self.lock.withLock {
        self.restoreContinuations.removeValue(forKey: requestId)
      }
      continuation?.resume(returning: .failed(self.bridgeError("restore_timeout")))
    }
  }

  private func purchaseResult(from payload: [String: Any]) -> PurchaseResult {
    switch (payload["type"] as? String)?.lowercased() {
    case "purchased": .purchased
    case "cancelled": .cancelled
    case "pending": .pending
    default: .failed(bridgeError((payload["message"] as? String) ?? "purchase_failed"))
    }
  }

  private func restoreResult(from payload: [String: Any]) -> RestoreResult {
    switch (payload["type"] as? String)?.lowercased() {
    case "restored": .restored
    case "no_purchases": .noPurchases
    default: .failed(bridgeError((payload["message"] as? String) ?? "restore_failed"))
    }
  }

  private func bridgeError(_ message: String) -> Error {
    NSError(
      domain: "io.nuxie.reactnative",
      code: 1,
      userInfo: [NSLocalizedDescriptionKey: message]
    )
  }
}
#endif
